DataDog/dd-trace-py
# feat(sci): add git metadata tags support

## Description
Add four environment variables:

- DD_GIT_REPOSITORY_URL string, default: empty
- DD_GIT_COMMIT_SHA string, default: empty
- DD_TRACE_GIT_METADATA_ENABLED bool, default: true
- DD_MAIN_PACKAGE string, default empty

Add following steps:

- Check if DD_TRACE_GIT_METADATA_ENABLED is True
- Check if git metadata is already provided by environment variables:
  - Check DD_GIT_REPOSITORY_URL and DD_GIT_COMMIT_SHA, if they exist, use metadata from them.
  - Check DD_TAGS, if it contains git.repository_url and git.commit.sha, use metadata from them.
  - Continue with metadata from artifacts.
- Read package name from DD_MAIN_PACKAGE, if it empty try detect package (will work for direct import)
- Read git metadata from the package metadata field Project-URL: source_code_link, split the value by "#" and use first part as repository URL and second part as commit hash

During trace telemetry sending:

- Add metadata tags to the first span of each chunk (NB: if [V7](https://github.com/DataDog/datadog-agent/blob/3ce2edbd37f3815102e31f59926699930bacdfc2/pkg/trace/api/api.go#L439) payload support already implemented send tags on payload level):
  - _dd.git.repository_url for repository URL
  - _dd.git.commit.sha for commit hash

During profile telemetry sending:
- Add metadata tags to each profile:
  - git.repository_url for repository URL
  - git.commit.sha to commit hash

## Checklist
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.





## Motivation
Add new git metadata tags and extraction support
RFC: [SCI Embedding Git references in built artifacts RFC](https://docs.google.com/document/d/1vxuRUNzHqd6sp1lnF3T383acbLrG0R-xDGi8cdZ4cs8/edit#)
Jira: [SC-1159](https://datadoghq.atlassian.net/browse/SC-1159)

## Design 


## Testing strategy
1. Build app from the cloned git repo with tracer/profiler with ddgitmetadata module in setup.py
1. Check git.repository_url and git.commit.sha tags on traces/profiles DD pages
1. Same as 1, but with DD_GIT_REPOSITORY_URL and DD_GIT_COMMIT_SHA with different values (values from env variables must present)
1. Same as 1, but with DD_TAGS="git.repository_url:some_repo git.commit.sha:some_sha" (values from env variables must present)
1. Same as 1, but with DD_TRACE_GIT_METADATA_ENABLED=false, tags must not be presented




## Relevant issue(s)


## Testing strategy




## Reviewer Checklist


[SC-1159]: https://datadoghq.atlassian.net/browse/SC-1159?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
